surfacers/cloudwatch/cloudwatch.go

```diff
@@ -0,0 +1,245 @@
+
+// The dimension named used to identify distributions
+const distributionDimensionName string = "le"
+
+type CWSurfacer struct {
+	c                   *configpb.SurfacerConf
+	writeChan           chan *metrics.EventMetrics
+	session             *cloudwatch.CloudWatch
+	l                   *logger.Logger
+	allowedMetricsRegex *regexp.Regexp
+
+	// A cache of []*cloudwatch.MetricDatum's, used for batch writing to the
+	// cloudwatch api.
+	cwMetricDatumCache []*cloudwatch.MetricDatum
+}
+
+func (cw *CWSurfacer) receiveMetricsFromEvent(ctx context.Context) {
+RoutineLoop:
+	for {
+		select {
+		case <-ctx.Done():
+			cw.l.Infof("Context canceled, stopping the surfacer write loop")
+			return
+		case em := <-cw.writeChan:
+
+			// evaluate if any of the metric labels are to be ignored and exit early
+			for _, label := range em.LabelsKeys() {
+				if cw.ignoreMetric(label) || cw.ignoreMetric(em.Label(label)) {
+					continue RoutineLoop
+				}
+			}
+
+			// check if a failure metric can be calculated
+			if em.Metric("success") != nil && em.Metric("total") != nil && em.Metric("failure") == nil {
+				if failure, err := calculateFailureMetric(em); err == nil {
+					em.AddMetric("failure", metrics.NewFloat(failure))
+				} else {
+					cw.l.Errorf("Error calculating failure metric: %s", err)
+				}
+			}
+
+			cw.recordEventMetrics(em)
+		}
+	}
+}
+
+// recordEventMetrics takes an EventMetric, which can contain multiple metrics of varying types, and loops through
+// each metric in the EventMetric, parsing each metric into a structure that is supported by Cloudwatch
+func (cw *CWSurfacer) recordEventMetrics(em *metrics.EventMetrics) {
+	for _, metricKey := range em.MetricsKeys() {
+
+		switch value := em.Metric(metricKey).(type) {
+		case metrics.NumValue:
+			cw.publishMetrics(cw.newCWMetricDatum(metricKey, value.Float64(), emLabelsToDimensions(em), em.Timestamp, em.LatencyUnit))
+
+		case *metrics.Map:
+			for _, mapKey := range value.Keys() {
+				dimensions := emLabelsToDimensions(em)
+				dimensions = append(dimensions, &cloudwatch.Dimension{
+					Name:  aws.String(value.MapName),
+					Value: aws.String(mapKey),
+				})
+				cw.publishMetrics(cw.newCWMetricDatum(metricKey, value.GetKey(mapKey).Float64(), dimensions, em.Timestamp, em.LatencyUnit))
+			}
+
+		case *metrics.Distribution:
+			for i, distributionBound := range value.Data().LowerBounds {
+				dimensions := append(emLabelsToDimensions(em), &cloudwatch.Dimension{
+					Name:  aws.String(distributionDimensionName),
+					Value: aws.String(strconv.FormatFloat(distributionBound, 'f', -1, 64)),
+				})
+
+				cw.publishMetrics(cw.newCWMetricDatum(metricKey, float64(value.Data().BucketCounts[i]), dimensions, em.Timestamp, em.LatencyUnit))
+			}
+		}
+	}
+}
+
+// publish the metrics to cloudwatch, using the namespace provided from configuration
+func (cw *CWSurfacer) publishMetrics(md *cloudwatch.MetricDatum) {
+	if len(cw.cwMetricDatumCache) >= 20 {
```

Comment:
You probably want to use cloudwatchMaxMetricDatums here instead of 20.

Reply:
sorry, that's an obvious miss by me - but thanks for catching

Reply:
No worries. :)